### PR TITLE
[FLINK-14416][table] Add Module interface and ModuleManager

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

Add Module interface and ModuleManager to Flink SQL.

## Brief change log

- developed Module interface and ModuleManager class
- added ModuleManager to impls of TableEnvironment and FunctionCatalog
- added module related APIs to TableEnvironment
- adapted lots of code (prod and testing) to adopt ModuleManager
- excluded module loading/unloading API from python, since flink python also doesn't allow registering catalogs

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

Docs will be added later

### CODE REVIEW

**Comment by bowenli86** [comment]
cc @xuefuz @lirui-apache @zjuwangg

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/internal/StreamTableEnvironmentImpl.java [hunk 1]
```diff
@@ -104,7 +106,9 @@ public static StreamTableEnvironment create(
 			settings.getBuiltInCatalogName(),
 			new GenericInMemoryCatalog(settings.getBuiltInCatalogName(), settings.getBuiltInDatabaseName()));
 
-		FunctionCatalog functionCatalog = new FunctionCatalog(catalogManager);
+		ModuleManager moduleManager = new ModuleManager();
+
+		FunctionCatalog functionCatalog = new FunctionCatalog(catalogManager, moduleManager);
```
Comment (xuefuz) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/internal/StreamTableEnvironmentImpl.java]: I saw repeated pattern like this. It's probably better to have a constructor that doesn't take a module manager and instead creates it inside the constructor.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java [hunk 2]
```diff
@@ -103,6 +106,23 @@ static TableEnvironment create(EnvironmentSettings settings) {
 	 */
 	Optional<Catalog> getCatalog(String catalogName);
 
+	/**
+	 * Load a module with given name.
```
Comment (xuefuz) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: Nit: use a similar comment as those in existing methods.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/internal/StreamTableEnvironmentImpl.java (hunk 1 above)
Comment (bowenli86) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/internal/StreamTableEnvironmentImpl.java]: TableEnv needs to get a hold on module manager to be able to load/unload module

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java [hunk 3]
```diff
@@ -237,6 +257,13 @@ static TableEnvironment create(EnvironmentSettings settings) {
 	 */
 	String[] listCatalogs();
 
+	/**
+	 * Gets the names of all modules registered in this environment.
+	 *
+	 * @return A list of the names of all registered modules.
+	 */
+	String[] listModules();
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: I will mention that in the comment

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 4]
```diff
@@ -118,7 +125,8 @@ public static TableEnvironmentImpl create(EnvironmentSettings settings) {
 			settings.getBuiltInCatalogName(),
 			new GenericInMemoryCatalog(settings.getBuiltInCatalogName(), settings.getBuiltInDatabaseName()));
 
-		FunctionCatalog functionCatalog = new FunctionCatalog(catalogManager);
+		ModuleManager moduleManager = new ModuleManager();
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: the constructor is protected and this is the static util calls the constructor

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	/**
+	 * Get an optional of {@link FunctionDefinition} by a given name.
+	 *
+	 * @param name name of the function
+	 * @return an optional of {@link FunctionDefinition}
+	 */
+	public Optional<FunctionDefinition> getFunctionDefinition(String name) {
+		Optional<Module> module = modules.values().stream()
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: we are good here since it's a stream of modules, not a stream of functions

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/module/ModuleManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.module;
+
+/**
+ * Test for {@link ModuleManager}.
+ */
+public class ModuleManagerTest {
+
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/module/ModuleManagerTest.java]: deleted for now

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java [hunk 7]
```diff
@@ -103,6 +106,23 @@ static TableEnvironment create(EnvironmentSettings settings) {
 	 */
 	Optional<Catalog> getCatalog(String catalogName);
 
+	/**
+	 * Load a {@link Module} under a unique name.
```
Comment (lirui-apache) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: mention that the module will be added to the end of list?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.functions.FunctionDefinition;
+
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Modules define a set of metadata, including functions, user defined types, operators, rules, etc.
+ * Metadata from modules are regarded as extensions of core built-in system that users can take advantages of.
```
Comment (lirui-apache) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java]: According to the design doc, core module may not be loaded if users specify custom modules. So in that sense modules are more than "extensions" of core built-in system -- they can completely replace the built-in system, right?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.functions.FunctionDefinition;
+
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Modules define a set of metadata, including functions, user defined types, operators, rules, etc.
```
Comment (lirui-apache) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java]: What's the implication if a custom module provides user defined types and what does that have to do with Catalogs? For example, if a custom module provides a custom TIMESTAMP type different from Flink core module's, Catalogs need to be aware of it so that they can return the proper CatalogTable based on this custom TIMESTAMP type?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	 */
+	public Set<String> listFunctions() {
+		return modules.values().stream()
+				.map(m -> m.listFunctions())
+				.flatMap(n -> n.stream())
+				.collect(Collectors.toSet());
+	}
+
+	/**
+	 * Get an optional of {@link FunctionDefinition} by a given name.
```
Comment (lirui-apache) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: mention how resolution is done if multiple modules have functions under this given name?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/internal/StreamTableEnvironmentImpl.java (hunk 1 above)
Comment (xuefuz) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/internal/StreamTableEnvironmentImpl.java]: Got it.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java [hunk 11]
```diff
@@ -103,6 +106,23 @@ static TableEnvironment create(EnvironmentSettings settings) {
...
+	 *
+	 * @param moduleName name of the {@link Module}
+	 * @param module the module instance
+	 * @throws ModuleAlreadyExistException thrown when there is already a module with the same name
```
Comment (xuefuz) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: Please make the comments for loadModule() and unloadModule() in a similar style as existing methods. For instance, Load => Loads.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (xuefuz) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: okay!

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	public void unloadModule(String name) throws ModuleNotFoundException {
+		if (modules.containsKey(name)) {
+			modules.remove(name);
+		} else {
+			throw new ModuleNotFoundException(name);
+		}
+	}
+
+	/**
+	 * Get names of all modules registered.
```
Comment (xuefuz) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: Nit: registered -> loaded

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (xuefuz) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: okay.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java (hunk 7 above)
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: mentioning "list" may be revealing too much implementation details on API. I'll rephrase the comment here

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java (hunk 8 above)
Comment (bowenli86) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java]: let me rephrase it

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java (hunk 9 above)
Comment (bowenli86) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java]: that hasn't been thought thru yet. my intuition is that catalog can be selective on which user defined types they support, it will needs its own FLIP in the future

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+				.collect(Collectors.toSet());
+	}
+
+	/**
+	 * Get an optional of {@link FunctionDefinition} by a given name.
+	 *
+	 * @param name name of the function
+	 * @return an optional of {@link FunctionDefinition}
+	 */
+	public Optional<FunctionDefinition> lookupFunctionDefinition(String name) {
```
Comment (xuefuz) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: Since "get" is used for module, I think it's good to be consistent. I'm fine with get.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java (hunk 11 above)
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: I see what you mean now with the example :)

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java (hunk 13 above)
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: ok

**Comment by bowenli86** [comment]
@xuefuz @lirui-apache can you take another look?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java (hunk 2 above)
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: Please extend the javadoc with description what is a `Module`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/exceptions/ModuleAlreadyExistException.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.module.exceptions;
+
+/**
+ * Exception for trying to load a module that already exists.
+ */
+public class ModuleAlreadyExistException extends Exception {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/exceptions/ModuleAlreadyExistException.java]: I would be strongly in favour of removing this exception, see my comment on `TableEnvironment`.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/exceptions/ModuleNotFoundException.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.module.exceptions;
+
+/**
+ * Exception for operating on a module that does not exists.
+ */
+public class ModuleNotFoundException extends Exception {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/exceptions/ModuleNotFoundException.java]: I would be strongly in favour of removing this exception, see my comment on `TableEnvironment`.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+	/**
+	 * Get names of all functions from all modules.
+	 *
+	 * @return a set of names of registered modules.
+	 */
+	public Set<String> listFunctions() {
+		return modules.values().stream()
+				.map(m -> m.listFunctions())
+				.flatMap(n -> n.stream())
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: Or remove one step: `.flatMap(m -> m.listFunctions().stream())`

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: Shouldn't the names be sorted?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+			throw new ModuleNotFoundException(name);
+		}
+	}
+
+	/**
+	 * Get names of all modules loaded.
+	 *
+	 * @return a list of names of modules loaded
+	 */
+	public List<String> listModules() {
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: Why is this not a set?

Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: the order of modules matters, to show the loaded order

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java [hunk 18]
```diff
@@ -103,6 +106,23 @@ static TableEnvironment create(EnvironmentSettings settings) {
 	 */
 	Optional<Catalog> getCatalog(String catalogName);
 
+	/**
+	 * Loads a {@link Module} under a unique name. Modules will be kept in the loaded order.
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: do we need to do that considering Module has its own javadoc?

APIs like `tEnv#registerCatalog()/registerFunction()` don't explain what a catalog/function is

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java [hunk 19]
```diff
@@ -103,6 +106,23 @@ static TableEnvironment create(EnvironmentSettings settings) {
...
+	 *
+	 * @param moduleName name of the {@link Module}
+	 * @param module the module instance
+	 * @throws ModuleAlreadyExistException thrown when there is already a module with the same name
+	 */
+	void loadModule(String moduleName, Module module) throws ModuleAlreadyExistException;
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: Sorry, that's a point I missed considering in the FLIP but realized when implementing it.

After comparing with how other APIs, like catalog in TableEnv, handle exceptions, I agree we should remove the checked exception. Similar to `CatalogException`, how about we create a `ModuleException extends RuntimeException` and it will be used very similar to `CatalogException`?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+	/**
+	 * Load a module under a unique name. Modules will be kept in the loaded order, and new module
+	 * will be added to the end.
+	 *
+	 * @param name name of the module
+	 * @param module the module instance
+	 * @throws ModuleAlreadyExistException thrown when there is already a module with the same name
+	 */
+	public void loadModule(String name, Module module) throws ModuleAlreadyExistException {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(name), "name cannot be null or empty string");
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: how about adopting how catalog works?

E.g. in CatalogManager, we have
```
public void registerCatalog(String catalogName, Catalog catalog) {
		checkArgument(!StringUtils.isNullOrWhitespaceOnly(catalogName), "Catalog name cannot be null or empty.");
		checkNotNull(catalog, "Catalog cannot be null");

		if (catalogs.containsKey(catalogName)) {
			throw new CatalogException(format("Catalog %s already exists.", catalogName));
		}
		...
	}
```

In ModuleManager, we adopt the same style as 

```
public void loadModule(String name, Module module) throws ModuleAlreadyExistException {
		checkArgument(!StringUtils.isNullOrWhitespaceOnly(name), "name cannot be null or empty string");
		checkNotNull(module, "module cannot be null");

		if (!modules.containsKey(name)) {
			modules.put(name, module);
		} else {
			throw new ModuleException("Module xxx already exist");
		}
	}

class ModuleException extends RuntimeException {}
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: why do function names in a module need to be sorted?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: You are right. Though `Catalog`s\`Function`s are rather a common term, whereas `Module`s are way broader. Personally, if I were the user, I would prefer to have a short information what can I do with the modules. It does not have to be in-depth, but sth along the lines, that 
`
Modules can provide system level extensions such as sytem functions, operators etc.
`

But feel free to ignore the comment.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java]: Do we really need a specific exception for it? Isn't `ValidationException` enough? What benefit do we get with a `ModuleException`? The `ValidationException`/`TableException` clearly separates user-side errors from system-side errors. If you really want to introduce additional exception let's try to maintain this distinction.

BTW, regarding the exceptions in Catalog API, I had a few discussions with @twalthr before and we both agree that at some point we should decrease the number of exceptions being thrown from the Catalog API. As of now it is not a very user friendly API in that regard. Therefore I would not take it as an example.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: I think that's how `SHOW <whatever>` works in most of the systems. It also makes sense as it is easier for users to look for his/hers objects.

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: After a second thought, this is not a user facing API though, we can still have a presentation layer on top, where we can sort it.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java (hunk 17 above)
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: Got it.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java]: yes, I agree.

**Comment by bowenli86** [comment]
feel we are losing track of comments, let's discuss them on slack @dawidwys
